web3/main.py

```diff
@@ -130,8 +136,12 @@ def __init__(self, provider=None, middlewares=None, modules=None, ens=empty):
         if modules is None:
             modules = get_default_modules()
 
-        for module_name, module_class in modules.items():
-            module_class.attach(self, module_name)
+        for module in modules:
+            module_class, *submodules = modules[module]
+            module_class.attach(self, module)
+            if submodules:
+                for subname, subclasses in submodules[0].items():
```

Comment:
Yeah, this `submodules[0]` looks surprising, I think. Also, if we're going to go with a recursive definition, might as well implement with a recursive solution, so it can go arbitrarily (well stack-limit) deep.

Something like a new utility method:
```py
    def attach_modules(parent_module, module_definitions)
       for module_name, module_info in module_definitions.items():
            module_class = module_info[0]
            module_class.attach(parent_module, module_name)

            if len(module_info) == 2:
                submodule_definitions = module_info[1]
                module = getattr(parent_module, module_name)
                attach_modules(module, submodule_definitions)
            elif len(module_info) != 1:
                raise ValidationError("Module definition can only have 1 or 2 elements")
```

Then the web3 init looks like:
```py
        Web3.attach_modules(self, modules)
```